pkg/chunk/cache/diskcache.go

```diff
@@ -116,25 +160,42 @@ func (d *Diskcache) fetch(key string) ([]byte, bool) {
 
 // StoreChunk puts a chunk into the cache.
 func (d *Diskcache) StoreChunk(ctx context.Context, key string, value []byte) error {
-	d.mtx.Lock()
-	defer d.mtx.Unlock()
-
 	bucket := hash(key) % d.buckets
+
+	mutex := bucket % numMutexes // Get the index of the mutex associated with this bucket
+	d.indexMutexes[mutex].Lock()
+	defer d.indexMutexes[mutex].Unlock()
+
+	if d.index[bucket] == key { // If chunk is already cached return nil
+		return nil
+	}
+
+	if d.index[bucket] == "" {
+		bucketsInitialized.Inc()
+	} else {
+		collisionsTotal.Inc()
+	}
+
 	buf := d.buf[bucket*bucketSize : (bucket+1)*bucketSize]
 
 	n, err := put([]byte(key), buf, 0)
 	if err != nil {
+		d.index[bucket] = ""
 		return err
 	}
 
 	_, err = put(value, buf, n)
 	if err != nil {
+		d.index[bucket] = ""
```

Comment:
Not a review comment, but just curious what the error might have been.

Reply:
Might even we worth logging?

Reply:
The error is currently returned and logged in the background cache. It only really happens if a chunk is too large for a bucket. That's a pretty rare occurrence I'd imagine, but it could happen one day.